dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,207 @@
+                )
+
+        # Supress import error on GoogleAuth warning
+        logging.getLogger("googleapiclient.discovery_cache").setLevel(
+            logging.ERROR
+        )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        gauth = GoogleAuth(settings_file=self.gdrive_credentials_path)
+        gauth.CommandLineAuth()
+        return GoogleDrive(gauth)
+
+    @property
+    @ratelimit.sleep_and_retry
+    @ratelimit.limits(calls=10, period=1)
+    def drive(self):
+        return self.raw_drive
+
+    def create_drive_item(self, parent_id, title):
+        item = self.drive.CreateFile(
+            {
+                "title": title,
+                "parents": [{"id": parent_id}],
+                "mimeType": self.FOLDER_MIME_TYPE,
+            }
+        )
+        item.Upload()
+        return item
+
+    def get_drive_item(self, name, parent_id):
+        return next(
+            iter(
+                self.drive.ListFile(
+                    {
+                        "q": "'%s' in parents and trashed=false and title='%s'"
+                        % (parent_id, name)
+                    }
+                ).GetList()
+            ),
+            None,
+        )
+
+    def resolve_remote_file(self, parent_id, path_parts, create):
+        for path_part in path_parts:
+            item = self.get_drive_item(path_part, parent_id)
+            if not item and create:
+                item = self.create_drive_item(parent_id, path_part)
+            elif not item:
+                return None
+            parent_id = item["id"]
+        return item
+
+    def get_path_id_from_cache(self, path_info):
+        file_id = ""
+        parts = path_info.path.split("/")
+        if parts and (parts[0] in self.cached_root_dirs):
+            parent_id = self.cached_root_dirs[parts[0]]
+            file_id = self.cached_root_dirs[parts[0]]
+            parts.pop(0)
+        else:
+            parent_id = path_info.netloc
+        return file_id, parent_id, parts
+
+    def get_path_id(self, path_info, create=False):
+        file_id, parent_id, parts = self.get_path_id_from_cache(path_info)
+
+        if not parts and file_id:
+            return file_id
+
+        file1 = self.resolve_remote_file(parent_id, parts, create)
+        return file1["id"] if file1 else ""
+
+    def exists(self, path_info):
+        return self.get_path_id(path_info) != ""
+
+    def batch_exists(self, path_infos, callback):
+        results = []
+        for path_info in path_infos:
+            results.append(self.exists(path_info))
+            callback.update(str(path_info))
+        return results
+
+    def _upload(self, from_file, to_info, name, no_progress_bar):
+        dirname = to_info.parent
+        if dirname:
+            parent_id = self.get_path_id(dirname, True)
+        else:
+            parent_id = to_info.netloc
+
+        file1 = self.drive.CreateFile(
```

Comment:
Does it actually create the file, so we should we now delete file if upload fails?

Reply:
It creates local instance of file object. Creation of file at remote disk happens with `file1.Upload()` call.
I don't think that anything will be created at remote disk if `Upload` fails.